### PR TITLE
apply sub-repo permissions filtering to git.BlameFile

### PR DESCRIPTION
**Pull request by mollylogue**
Check if user has access to the file in `git.BlameFile` via `SubRepoPermissionChecker`

https://github.com/sourcegraph/sourcegraph/issues/28918

### CODE REVIEW

File: cmd/frontend/graphqlbackend/git_blob.go [hunk 1]
```diff
@@ -16,7 +17,7 @@ func (r *GitTreeEntryResolver) Blame(ctx context.Context,
 		NewestCommit: api.CommitID(r.commit.OID()),
 		StartLine:    int(args.StartLine),
 		EndLine:      int(args.EndLine),
-	})
+	}, authz.DefaultSubRepoPermsChecker)
```
Comment (mollylogue) [inline cmd/frontend/graphqlbackend/git_blob.go:20]: This is initialized at the top level in [cmd/frontend/internal/cli/serve_cmd.go](https://github.com/sourcegraph/sourcegraph/pull/28859/https://github.com/sourcegraph/sourcegraph/pull/28859/files#diff-5555619d23aebc262d7144fba501ab3cef27b978c48224c690bac3221771b3d0)

**Comment by mollylogue** [comment]
Outstanding questions:
- How can I test this from the top level? Do we have a test environment/code host w/ sub-repo permissions that I could leverage to do that?

File: internal/vcs/git/blame.go [hunk 2]
```diff
@@ -8,6 +8,10 @@ import (
 	"strings"
 	"time"
 
+	"github.com/sourcegraph/sourcegraph/internal/actor"
```
Comment (ryanslade) [inline internal/vcs/git/blame.go:11]: The usual 😄 

We need to find a way to make this work.. between just the three of us using GoLand on this team it adds up to quite a lot of wasted time

File: internal/vcs/git/blame_test.go [hunk 3]
```diff
@@ -6,6 +6,10 @@ import (
 	"testing"
 	"time"
 
+	"github.com/sourcegraph/sourcegraph/internal/actor"
```
Comment (ryanslade) [inline internal/vcs/git/blame_test.go:9]: Imports 😄

File: internal/vcs/git/blame_test.go [hunk 4]
```diff
@@ -69,14 +78,38 @@ func TestRepository_BlameFile(t *testing.T) {
...
+		// Case: user doesn't have access to the file, nothing returned.
+		checker.PermissionsFunc.SetDefaultHook(func(ctx context.Context, i int32, content authz.RepoContent) (authz.Perms, error) {
+			return authz.None, nil
+		})
+		runBlameFileTest(ctx, t, test.repo, test.path, test.opt, checker, label, nil)
+	}
+}
+
+func runBlameFileTest(ctx context.Context, t *testing.T, repo api.RepoName, path string, opt *BlameOptions,
+	checker authz.SubRepoPermissionChecker, label string, wantHunks []*Hunk) {
```
Comment (ryanslade) [inline internal/vcs/git/blame_test.go:106]: Could be worth adding `t.Helper()` at the start of this function. That way we'll know more easily which of the calls to `runBlameTest` failed.

File: internal/vcs/git/blame.go [hunk 5]
```diff
@@ -37,16 +41,27 @@ type Hunk struct {
...
 	span.SetTag("path", path)
 	span.SetTag("opt", opt)
 	defer span.Finish()
-	return blameFileCmd(ctx, gitserverCmdFunc(repo), path, opt)
+	return blameFileCmd(ctx, gitserverCmdFunc(repo), path, opt, repo, checker)
 }
 
-func blameFileCmd(ctx context.Context, command cmdFunc, path string, opt *BlameOptions) ([]*Hunk, error) {
+func blameFileCmd(ctx context.Context, command cmdFunc, path string, opt *BlameOptions, repo api.RepoName, checker authz.SubRepoPermissionChecker) ([]*Hunk, error) {
+	if checker.Enabled() {
```
Comment (ryanslade) [inline internal/vcs/git/blame.go:54]: Just to be extra safe we should also make sure that checker isn't nil

`if checker != nil && checker.Enabled()`

**Review by ryanslade — APPROVED** [review]
Nice!

File: internal/vcs/git/blame.go (hunk 2 above)
Comment (mollylogue) [reply internal/vcs/git/blame.go:11]: Noooo not the imports again 😣  `goimports` hates me.

File: internal/vcs/git/blame_test.go (hunk 4 above)
Comment (mollylogue) [reply internal/vcs/git/blame_test.go:106]: 👍 Sounds good. I'm so used to using the `ginkgo/gomega` testing ecosystem so I'm still adjusting to using the normal `testing` libraries

File: internal/vcs/git/blame.go (hunk 5 above)
Comment (mollylogue) [reply internal/vcs/git/blame.go:54]: 👍

**Pull request merged**